src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -648,13 +648,11 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
         case BOOL_TAG   => BooleanTpe
         case 'L' =>
           def processInner(tp: Type): Type = tp match {
-            case TypeRef(pre, sym, args) if (!sym.isStatic) =>
-              typeRef(processInner(pre.widen), sym, args)
-            case _ =>
-              tp
+            case TypeRef(_pre, _sym, _args) if !_sym.isStatic => typeRef(processInner(_pre.widen), _sym, _args)
```

Comment:
Hmm... Seeing diffs like this makes me wonder. Naming things is already absorbing a lot of time when coding and the new lint will intensify that. People will then resort to fixes like this one, making the code look worse.

I think this is a good example: writing `case TypeRef(pre, sym, args)` is very common in this codebase and everybody expects it to bind new names. If we can no longer write this it's distracting.

Here's an example which you pointed out as a good one (https://github.com/scala/scala/pull/8806/files#r1428583071):

<img width="568" alt="image" src="https://github.com/scala/scala/assets/119636/b20bbe68-e7b0-4652-b57d-ac92fe013a34">

I agree this is a good change.

Unfortunately I don't have an idea right now to tell the good and worse cases apart...

Reply:
💡 ⚡ it occurs to me that you must allow canonical names as determined by constructor pattern (case params) or extractor (unapply params).  Then we're not just looking at the binding but the context. I'll give that a shot.

However, I did convince myself that the shadowing problem is real. What is `tree`? That's more urgent when `fun` gets recycled a few times. My provisional fix was to stop inventing names and use `_foo` to mean a kind of "pass-through", if it's used immediately in RHS expr and not referenced 100 LOC away. There was one `classSym` like that, not sure I addressed that particular one.

The motivating case is patvar shadowing other kinds of definitions in scope, so maybe that is another dimension.

Also on naming, it wasn't always immediately obvious what a "good name" would be without a bit of study (which I avoided for reasons of scope).

My erroneous refactor shows how treacherous it can be. (I did not use intellij to perform renames, so maybe it falls under "things you can't do without an IDE".)

Another open question is the form `f(x) match { case x => }`, where x is an input to the scrutinee.